### PR TITLE
Replace most uses of say() on tape recorder with balloon alerts to reduce spam

### PR DESCRIPTION
**Pull request by cacogen**
## About The Pull Request
The tape recorder would spam the chat with multiple spoken lines when Play was pressed on an empty tape or when playback ended. This fixes that by replacing the spoken lines with balloon alerts only visible to the user of the tape recorder.

Playback still uses say().
## Why It's Good For The Game
Less chat spam
## Changelog
:cl:
qol: Tape recorder actions (e.g. starting/stopping playback) now use balloon alerts instead of say() to reduce chat spam
fix: You can no longer get a tape stuck in the tape recorder by unspooling it before inserting
/:cl:

### CODE REVIEW

**Comment by Iamgoofball** [comment]
this makes it impossible to play a tape recorder into a radio :-1:

File: code/game/objects/items/devices/taperecorder.dm [hunk 1]
```diff
@@ -91,19 +91,24 @@
...
 /obj/item/taperecorder/proc/eject(mob/user)
-	if(mytape)
-		playsound(src, 'sound/items/taperecorder/taperecorder_open.ogg', 50, FALSE)
-		to_chat(user, span_notice("You remove [mytape] from [src]."))
-		stop()
-		user.put_in_hands(mytape)
-		mytape = null
-		update_appearance()
+	if(!mytape)
+		balloon_alert_to_viewers("no tape!", vision_distance = 1)
```
Comment (Mothblocks) [inline code/game/objects/items/devices/taperecorder.dm:101]:
```suggestion
		balloon_alert(user, "no tape!")
```

File: code/game/objects/items/devices/taperecorder.dm [hunk 2]
```diff
@@ -91,19 +91,24 @@
...
+		return
+	if(playing)
+		balloon_alert_to_viewers("stop the tape first!", vision_distance = 1)
```
Comment (Mothblocks) [inline code/game/objects/items/devices/taperecorder.dm:104]:
```suggestion
		balloon_alert(user, "stop the tape first!")
```

File: code/game/objects/items/devices/taperecorder.dm [hunk 3]
```diff
@@ -91,19 +91,24 @@
...
+		return
+	playsound(src, 'sound/items/taperecorder/taperecorder_open.ogg', 50, FALSE)
+	balloon_alert_to_viewers("ejected [mytape]!", vision_distance = 1)
```
Comment (Mothblocks) [inline code/game/objects/items/devices/taperecorder.dm:107]:
```suggestion
	balloon_alert(user, "ejected tape")
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
I'll review the rest later but make sure you follow those rules throughout everything else

**Comment by VladinXXV** [comment]
> Playback still uses say().

Playback still uses say(), according to the description, so I don't see how this changes?

**Comment by VladinXXV** [comment]
> this makes it impossible to play a tape recorder into a radio 👎

Playback still uses say(), according to the description, so I don't see how this changes?

**Comment by BlueMemesauce** [comment]
I see how it could be confusing if you only read the changelog. It should say something like qol: Tape recorder actions (starting/stopping playback) now use balloon alerts to reduce spam

**Comment by VladinXXV** [comment]
> I see how it could be confusing if you only read the changelog. It should say something like qol: Tape recorder actions (starting/stopping playback) now use balloon alerts to reduce spam

Good point, yeah, that changelog is confusing.

File: code/game/objects/items/devices/taperecorder.dm [hunk 4]
```diff
@@ -91,19 +91,24 @@
 		if(!user.transferItemToLoc(I,src))
 			return
 		mytape = I
-		to_chat(user, span_notice("You insert [I] into [src]."))
+		balloon_alert_to_viewers("inserted [mytape]!", vision_distance = 1)
```
[the remark this answers is not in the document]
Comment (cacogen) [reply code/game/objects/items/devices/taperecorder.dm:94]: Thanks

File: code/game/objects/items/devices/taperecorder.dm [hunk 5]
```diff
@@ -182,13 +193,14 @@
 			used += 1 SECONDS
 			if(max - used < time_left_warning && !time_warned)
 				time_warned = TRUE
-				say("[(max - used) / 10] seconds left!") //deciseconds / 10 = seconds
+				balloon_alert(usr, "[(max - used) / 10] seconds left")
```
Comment (Mothblocks) [inline code/game/objects/items/devices/taperecorder.dm:196]:
```suggestion
				balloon_alert(usr, "[(max - used) / 10] second\s left")
```

File: code/game/objects/items/devices/taperecorder.dm [hunk 6]
```diff
@@ -237,7 +257,8 @@
 		if(playing == FALSE)
 			break
 		if(mytape.storedinfo.len < i)
-			say("End of recording.")
+			balloon_alert(usr, "recording ended")
+			sleep(1 SECONDS) //prevents multiple balloon alerts covering each other
```
Comment (Mothblocks) [inline code/game/objects/items/devices/taperecorder.dm:261]:
```suggestion
			stoplag(1 SECONDS) //prevents multiple balloon alerts covering each other
```

File: code/game/objects/items/devices/taperecorder.dm [hunk 7]
```diff
@@ -256,10 +277,7 @@
 
 /obj/item/taperecorder/attack_self(mob/user)
 	if(!mytape)
-		to_chat(user, span_notice("\The [src] is empty."))
-		return
-	if(mytape.unspooled)
-		to_chat(user, span_warning("\The tape inside \the [src] is broken!"))
+		balloon_alert(user, "[src] is empty!")
```
Comment (Mothblocks) [inline code/game/objects/items/devices/taperecorder.dm:280]:
```suggestion
		balloon_alert(user, "it's empty!")
```

File: code/game/objects/items/devices/taperecorder.dm [hunk 8]
```diff
@@ -285,15 +303,22 @@
 
 	var/list/transcribed_info = mytape.storedinfo
 	if(!length(transcribed_info))
-		return
-	if(!mytape)
+		balloon_alert(usr, "[mytape] is empty!")
```
Comment (Mothblocks) [inline code/game/objects/items/devices/taperecorder.dm:306]:
```suggestion
		balloon_alert(usr, "tape is empty!")
```

File: code/game/objects/items/devices/taperecorder.dm [hunk 9]
```diff
@@ -328,7 +353,7 @@
 	transcript_paper.name = "[paper_name] page [page_count]"
 	transcript_paper.update_appearance()
 
-	say("Transcript printed, [page_count] pages.")
+	balloon_alert(usr, "transcript printed, [page_count] pages")
```
Comment (Mothblocks) [inline code/game/objects/items/devices/taperecorder.dm:356]:
```suggestion
	balloon_alert(usr, "transcript printed\n[page_count] page\s")
```

File: code/game/objects/items/devices/taperecorder.dm [hunk 10]
```diff
@@ -308,7 +333,7 @@
 
 		// Very unexpected. Better abort non-gracefully.
 		if(excerpt_length > MAX_PAPER_LENGTH)
-			say("Error: Data corruption detected. Cannot print.")
+			balloon_alert(usr, "data corrupted, cannot print!")
```
Comment (Mothblocks) [inline code/game/objects/items/devices/taperecorder.dm:336]:
```suggestion
			balloon_alert(usr, "data corrupted, can't print!")
```

File: code/game/objects/items/devices/taperecorder.dm [hunk 11]
```diff
@@ -409,13 +434,13 @@
 				if(loc != user)
 					return
 				tapeflip()
-				to_chat(user, span_notice("You turn \the [src] over."))
+				balloon_alert(user, "flipped [src]")
```
Comment (Mothblocks) [inline code/game/objects/items/devices/taperecorder.dm:437]:
```suggestion
				balloon_alert(user, "flipped tape")
```

File: code/game/objects/items/devices/taperecorder.dm [hunk 12]
```diff
@@ -409,13 +434,13 @@
...
 				playsound(src, 'sound/items/taperecorder/tape_flip.ogg', 70, FALSE)
 			if("Unwind tape")
 				if(loc != user)
 					return
 				unspool()
-				to_chat(user, span_warning("You pull out all the tape!"))
+				balloon_alert(user, "unspooled [src]")
```
Comment (Mothblocks) [inline code/game/objects/items/devices/taperecorder.dm:443]:
```suggestion
				balloon_alert(user, "unspooled tape")
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Pull request merged**